corgi/core/authentication.py

```diff
@@ -0,0 +1,76 @@
+from django.core.exceptions import ObjectDoesNotExist
+from mozilla_django_oidc.auth import OIDCAuthenticationBackend
+from rest_framework.permissions import BasePermission
+
+from corgi.core.models import RedHatUser
+
+
+class CorgiOIDCBackend(OIDCAuthenticationBackend):
+    """An extension of mozilla_django_oidc's authentication backend
+    which customizes user creation and authentication to support Red
+    Hat SSO additional claims."""
+
+    def verify_claims(self, claims):
+        """Require, at a minimum, that a user have a rhatUUID claim before even trying to
+        authenticate them."""
+        verified = super(CorgiOIDCBackend, self).verify_claims(claims)
+        return verified and "rhatUUID" in claims
+
+    def filter_user_by_claims(self, claims):
+        """The default behavior is to use e-mail, which may not be unique.
+        Instead, we use Red Hat UUID, which should be unique and persistent
+        between changes to other user claims."""
+        rhat_uuid = claims.get("rhatUUID")
+
+        if not rhat_uuid:
+            return self.UserModel.objects.none()
```

Comment:
I think mypy might have complained at you before because this returns an empty UserModel queryset here and in the below except block, but a RedHat user queryset (with a value) on success. See if below fixes those mypy errors and lets you add type hints to all your methods:
```suggestion
            return RedHatUser.objects.none()
```
Above also avoids needing the self arg, so this can become a static method. Or maybe we should always use self.UserModel, and the code on line 29 below that accesses RedHatUser directly should be changed?

Reply:
I may have chosen a poor name for `RedHatUser`, which is not actually a User object at all, just a separate object that stores additional information from Red Hat SSO. This method wants a queryset, but now that I look more closely at it, I realize I had a typo that could have been an issue. Anyway, I'll rename `RedHatUser` to `RedHatProfile` which should make the distinction clearer.

Reply:
The issue here is not so much the naming, but the return type. Because we return different queryset types on success vs. failure, mypy will complain.

I think you can fix this by always returning a RedHatUser.objects.none() on failure instead of self.UserModel.objects.none(). Even if the first is really a profile and not a user, I think it's OK since we're just returning an empty query set anyway.

Reply:
I added some type annotations here to hopefully clear things up a bit. The return type of this method is Iterable[User]. On line 34, the return is actually the User object referenced by the Profile, not the Profile itself.